HIVE-29122: Vectorization - Support IGNORE NULLS for FIRST_VALUE and …

soumyakanti3578 opened the pull request:
…LAST_VALUE

<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://cwiki.apache.org/confluence/display/Hive/HowToContribute
  2. Ensure that you have created an issue on the Hive project JIRA: https://issues.apache.org/jira/projects/HIVE/summary
  3. Ensure you have added or run the appropriate tests for your PR: 
  4. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP]HIVE-XXXXX:  Your PR title ...'.
  5. Be sure to keep the PR description updated to reflect all changes.
  6. Please write your PR title to summarize what this PR proposes.
  7. If possible, provide a concise example to reproduce the issue for a faster review.

-->

### What changes were proposed in this pull request?
<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If you fix some SQL features, you can provide some references of other DBMSes.
  3. If there is design documentation, please add the link.
  4. If there is a discussion in the mailing list, please add the link.
-->
Pass `respectNulls` to the `first_value` and `last_value` VectorPTFEvaluator classes and update `evaluateGroupBatch` methods to support `IGNORE NULLS`


### Why are the changes needed?
<!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, you can clarify why it is a bug.
-->
Explained in detail: https://issues.apache.org/jira/browse/HIVE-29122

### Does this PR introduce _any_ user-facing change?
<!--
Note that it means *any* user-facing change including all aspects such as the documentation fix.
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description, screenshot and/or a reproducable example to show the behavior difference if possible.
If possible, please also clarify if this is a user-facing change compared to the released Hive versions or within the unreleased branches such as master.
If no, write 'No'.
-->
No

### How was this patch tested?
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
-->
```
mvn test -pl itests/qtest -Pitests -Dtest=TestMiniLlapLocalCliDriver -Dtest.output.overwrite=true -Dqfile="vectorized_first_last_value_ignore_nulls_double.q,vectorized_first_last_value_ignore_nulls_int.q,vectorized_first_last_value_ignore_nulls_decimal.q"
```

soumyakanti3578 commented:
@kasakrisz 
I debugged a bit to understand the workflow. To simulate the scenario where rows with the same key doesn't fit into 1 single batch, we can run the tests with `SET hive.vectorized.testing.reducer.batch.size=2;` With this I have noticed that `VectorPTFGroupBatches.bufferedBatches` contains batches whose size is at most 2. 

For the data in `vectorized_first_last_value_ignore_nulls_int.q`, for key 1 and 2, I have seen batches like:
```
Batch 1:   [[1, null], [1, null]]
Batch 2:   [1, null]
Batch 3:   [[2, null], [2, 20]]
Batch 4:   [2, null]
```
Let's look at `last_value ignore nulls` computation.
First we will compute last value for Range with key = 1, which consists of Batch 1 and 2. A new evaluator will be created for `VectorPTFEvaluatorLongLastValue`, which we will keep reusing.

First, we will call `VectorPTFEvaluatorLongLastValue.evaluateGroupBatch(batch 1)` will be called, and we will have to check both vector elements for non null values. Since they are null, we will have `isGroupResultNull = true` and `lastValue = null`.

Then, we will call `VectorPTFEvaluatorLongLastValue.evaluateGroupBatch(batch 2)` on the same evaluator object, but the  value of `isGroupResultNull` and `lastValue` doesn't change.

Next we will call `VectorPTFEvaluatorLongLastValue.evaluateGroupBatch(batch 3)` on the same evaluator, and this time, the values will be `isGroupResultNull = false` and `lastValue = 20`.

And then, we will call `VectorPTFEvaluatorLongLastValue.evaluateGroupBatch(batch 4)` on the same evaluator, and the values will be `isGroupResultNull = true` and `lastValue = 20`. Note that `lastValue` remains unchanged.

Now if we didn't have other rows in the table, we would get out of the `while` loop in `VectorPTFGroupBatches.BlockIterator.run` and call `getEvaluatorResult(evaluator)` as seen below:
```
    Object run(VectorPTFEvaluatorBase evaluator) throws HiveException {
      runEvaluator(evaluator, range);

      while (hasNext()) {
        nextBlock();
        runEvaluator(evaluator, range);
      }

      return getEvaluatorResult(evaluator);
    }

    Object getEvaluatorResult(VectorPTFEvaluatorBase evaluator) {
      return evaluator.isGroupResultNull() ? null : evaluator.getGroupResult();
    }
```

`VectorPTFEvaluatorLongLastValue.isGroupResultNull()` was updated in this PR as:
```
  public boolean isGroupResultNull() {
    return isGroupResultNull && doesRespectNulls();
  }
```
This returns `false` even though `isGroupResultNull = true` at the end because `doesRespectNulls()` returns false. So finally `evaluator.getGroupResult()` is called, which returns `lastValue = 20`.

If we were running `last value respect nulls`, it would have returned `null` because `isGroupResultNull()` would have returned `true`.

Similarly `first value ignore nulls` can be explained too. One important difference is that `VectorPTFEvaluatorLongFirstValue` has the boolean variable `haveFirstValue`. Once we find a non null value, this variable is set to true and for all batches after that we directly write the value into the output vector.

Hopefully this makes sense, but please let me know if I am missing something. If needed, I can add `SET hive.vectorized.testing.reducer.batch.size=2;` to all three test files too - they are all passing on my local machine.

deniskuzZ commented:
> And then, we will call VectorPTFEvaluatorLongLastValue.evaluateGroupBatch(batch 4) on the same evaluator, and the values will be isGroupResultNull = true and lastValue = 20. Note that lastValue remains unchanged.

shouldn't the `isGroupResultNull` stay false?

soumyakanti3578 commented:
@deniskuzZ 
Yes, but it didn't affect the results. But in the latest commit I have updated the initial value to `isGroupResultNull = lastValue == null;` which should take care of the inconsistency :) 

@kasakrisz 
I have added more tests in all three files with `set hive.vectorized.testing.reducer.batch.size=2;`

Please review again! Thanks!

kasakrisz reviewed (approved):
+1 pending tests
